blocks/graph.py

```diff
@@ -161,8 +162,39 @@ def replace(self, replacements):
             substitutes.
 
         """
-        return ComputationGraph(theano.clone(self.outputs,
-                                             replace=replacements))
+        # Due to theano specifics we have to make one replacement in time
+        replacements = OrderedDict(replacements)
+
+        apply_nodes_sorted = theano.gof.graph.io_toposort(
+            self.inputs, self.outputs)
+        outputs_cur = self.outputs
+        # Sort `replacements` in topological order
+        replacements_sorted = []
+        for node in apply_nodes_sorted:
+            for input_ in node.inputs:
+                if input_ not in replacements:
+                    continue
+                replacements_sorted.append((input_, replacements[input_]))
+        # `replacements` with previous replacements applied. We have to track
+        # variables in the new graph corresponding to original replacements.
+        replacements_cur = replacements_sorted
```

Comment:
Now it looks correct to me. This iterative logic should also be documented, it would be nice to use your test as a doctest.

Suggestions:

- remove `replacements_sorted` because it is not used anywhere except this assignment

- use an `OrderedDict` for `replacements_cur` as you did. Why did you switch to a list of tuples?